### PR TITLE
[tools] added change registry script

### PR DESCRIPTION
**Pull request by RomanenkoDenys**
Signed-off-by: Denis Romanenko <denis.romanenko@flant.com>

## Description

Added change registry script to simplify change of registry for devops.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: tools
type: chore
summary: Added script to change container registry, which makes it easier for DevOps engineers.
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 1]
```diff
@@ -199,12 +199,9 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
 To switch the Deckhouse cluster to using a third-party registry, follow these steps:
 
 * Update the `image` field in the `d8-system/deckhouse` deployment to contain the address of the Deckhouse image in the third-party-registry;
-* Edit the `d8-system/deckhouse-registry` secret (note that all parameters are Base64-encoded):
-  * Insert third-party registry credentials into `.dockerconfigjson`.
-  * Replace `address` with the third-party registry's host address (e.g., `registry.example.com`).
-  * Change `path` to point to a repo in the third-party registry (e.g., `/deckhouse/ee`).
-  * If necessary, change `scheme` to `http` (if the third-party registry uses HTTP scheme).
-  * If necessary, change or add the `ca` field with the root CA certificate that validates the third-party registry's https certificate (if the third-party registry uses self-signed certificates).
+* Download script [change-registry.sh](https://github.com/deckhouse/deckhouse/blob/main/tools/change-registry.sh) to the master node and start them with third-party-registry parameters.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:202]:
```suggestion
* If the third-party-registry uses a self-signed certificate, put the root CA certificate that validates the third-party registry's HTTPS certificate to file `ca.crt` near the script and add the `--ca-file ca.crt` option to the script.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 2]
```diff
@@ -199,12 +199,9 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
...
+  * Example: ```./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password```.
+  * If the third-party-registry uses self-signed certificate, put root CA certificate that validates the third-party registry's https certificate to file `ca.crt` near script and add option ```--ca-file ca.crt``` to the script run.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:204]:
```suggestion
  * If the third-party-registry uses a self-signed certificate, put the root CA certificate that validates the third-party registry's HTTPS certificate to file `ca.crt` near the script and add the `--ca-file ca.crt` option to the script.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 3]
```diff
@@ -199,12 +199,9 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
...
 
 * Update the `image` field in the `d8-system/deckhouse` deployment to contain the address of the Deckhouse image in the third-party-registry;
-* Edit the `d8-system/deckhouse-registry` secret (note that all parameters are Base64-encoded):
-  * Insert third-party registry credentials into `.dockerconfigjson`.
-  * Replace `address` with the third-party registry's host address (e.g., `registry.example.com`).
-  * Change `path` to point to a repo in the third-party registry (e.g., `/deckhouse/ee`).
-  * If necessary, change `scheme` to `http` (if the third-party registry uses HTTP scheme).
-  * If necessary, change or add the `ca` field with the root CA certificate that validates the third-party registry's https certificate (if the third-party registry uses self-signed certificates).
+* Download script [change-registry.sh](https://github.com/deckhouse/deckhouse/blob/main/tools/change-registry.sh) to the master node and start them with third-party-registry parameters.
+  * Example: ```./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password```.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:203]:
```suggestion
  * Example:

        ```shell
    ./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password
    ```
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 4]
```diff
@@ -203,12 +203,9 @@ chmod 700 d8-push.sh
...
 
 * Измените поле `image` в Deployment `d8-system/deckhouse` на адрес образа Deckhouse в новом registry;
-* Измените Secret `d8-system/deckhouse-registry` (все параметры хранятся в кодировке Base64):
-  * Исправьте `.dockerconfigjson` с учетом авторизации в новом registry.
-  * Исправьте `address` на адрес нового registry (например, `registry.example.com`).
-  * Исправьте `path` на путь к репозиторию Deckhouse в новом registry (например, `/deckhouse/ee`).
-  * При необходимости измените `scheme` на `http` (если используется HTTP registry).
-  * Если registry использует самоподписные сертификаты, то измените или добавьте поле `ca`, куда внесите корневой сертификат соответствующего сертификата registry;
+* Скачайте скрипт [change-registry.sh](https://github.com/deckhouse/deckhouse/blob/main/tools/change-registry.sh) на мастер-узел и запустите его с параметрами нового registry.
+  * Пример запуска: ```./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password```.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:207]:
```suggestion
  * Пример запуска:
    ```
    ./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password
    ```
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 5]
```diff
@@ -203,12 +203,9 @@ chmod 700 d8-push.sh
...
+  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` возле скрипта и добавьте к вызову опцию ```--ca-file ca.crt```.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:208]:
```suggestion
  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` возле скрипта и добавьте к вызову опцию `--ca-file ca.crt`.
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 6]
```diff
@@ -199,12 +199,13 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
...
-  * Replace `address` with the third-party registry's host address (e.g., `registry.example.com`).
-  * Change `path` to point to a repo in the third-party registry (e.g., `/deckhouse/ee`).
-  * If necessary, change `scheme` to `http` (if the third-party registry uses HTTP scheme).
-  * If necessary, change or add the `ca` field with the root CA certificate that validates the third-party registry's https certificate (if the third-party registry uses self-signed certificates).
+* Download script [change-registry.sh](https://github.com/deckhouse/deckhouse/blob/main/tools/change-registry.sh) to the master node and start them with third-party-registry parameters.
+  * Example:
+
+    ```shell
+    ./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password
+    ```
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:207]:
```suggestion

    ```
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 7]
```diff
@@ -199,12 +199,13 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
...
+  * If the third-party-registry uses a self-signed certificate, put the root CA certificate that validates the third-party registry's HTTPS certificate to file `ca.crt` near the script and add the `--ca-file ca.crt` option to the script.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:208]:
```suggestion

  * If the third-party-registry uses a self-signed certificate, put the root CA certificate that validates the third-party registry's HTTPS certificate to file `ca.crt` near the script and add the `--ca-file ca.crt` option to the script.
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 8]
```diff
@@ -203,12 +203,13 @@ chmod 700 d8-push.sh
...
-* Измените Secret `d8-system/deckhouse-registry` (все параметры хранятся в кодировке Base64):
-  * Исправьте `.dockerconfigjson` с учетом авторизации в новом registry.
-  * Исправьте `address` на адрес нового registry (например, `registry.example.com`).
-  * Исправьте `path` на путь к репозиторию Deckhouse в новом registry (например, `/deckhouse/ee`).
-  * При необходимости измените `scheme` на `http` (если используется HTTP registry).
-  * Если registry использует самоподписные сертификаты, то измените или добавьте поле `ca`, куда внесите корневой сертификат соответствующего сертификата registry;
+* Скачайте скрипт [change-registry.sh](https://github.com/deckhouse/deckhouse/blob/main/tools/change-registry.sh) на мастер-узел и запустите его с параметрами нового registry.
+  * Пример запуска:
+
+    ```
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:209]:
```suggestion
    ```shell
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 9]
```diff
@@ -203,12 +203,13 @@ chmod 700 d8-push.sh
...
+    ./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password
+    ```
+  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` возле скрипта и добавьте к вызову опцию `--ca-file ca.crt`.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:212]:
```suggestion

  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` возле скрипта и добавьте к вызову опцию `--ca-file ca.crt`.
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

**Review by z9r5 — APPROVED** [review]

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 10]
```diff
@@ -199,12 +199,14 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
 To switch the Deckhouse cluster to using a third-party registry, follow these steps:
 
 * Update the `image` field in the `d8-system/deckhouse` deployment to contain the address of the Deckhouse image in the third-party-registry;
-* Edit the `d8-system/deckhouse-registry` secret (note that all parameters are Base64-encoded):
-  * Insert third-party registry credentials into `.dockerconfigjson`.
-  * Replace `address` with the third-party registry's host address (e.g., `registry.example.com`).
-  * Change `path` to point to a repo in the third-party registry (e.g., `/deckhouse/ee`).
-  * If necessary, change `scheme` to `http` (if the third-party registry uses HTTP scheme).
-  * If necessary, change or add the `ca` field with the root CA certificate that validates the third-party registry's https certificate (if the third-party registry uses self-signed certificates).
+* Download script [change-registry.sh](https://github.com/deckhouse/deckhouse/blob/main/tools/change-registry.sh) to the master node and start them with third-party-registry parameters.
```
Comment (konstantin-axenov) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:202]:
```suggestion
* Download script [change-registry.sh](https://github.com/deckhouse/deckhouse/blob/main/tools/change-registry.sh) to the master node and run it with parameters for a new registry.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 11]
```diff
@@ -199,12 +199,14 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
...
+  * Example:
+
+    ```shell
```
Comment (konstantin-axenov) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:205]: I suggest adding a command to download script like [here](https://github.com/deckhouse/deckhouse/blob/main/docs/documentation/pages/DECKHOUSE-FAQ.md?plain=1#L164).

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 12]
```diff
@@ -199,12 +199,14 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
...
+    ./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password
+    ```
+
+  * If the third-party-registry uses a self-signed certificate, put the root CA certificate that validates the third-party registry's HTTPS certificate to file `ca.crt` near the script and add the `--ca-file ca.crt` option to the script.
```
Comment (konstantin-axenov) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:209]:
```suggestion
  * If the registry uses a self-signed certificate, put the root CA certificate that validates the registry's HTTPS certificate to file `ca.crt` near the script and add the `--ca-file ca.crt` option to the script.
```
This script can be used to switch back to registry.deckhouse.io, so seems better not to use `third-party` term.

File: tools/change-registry.sh [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+DOCKERCONFIGJSON="$(create_dockerconfigjson | base64 -w0)"
+REGISTRY_ADDRESS="$(echo -n $REGISTRY_ADDRESS | base64 -w0)"
+REGISTRY_PATH="$(echo -n $REGISTRY_PATH | base64 -w0)"
+REGISTRY_SCHEME="$(echo -n $REGISTRY_SCHEME | base64 -w0)"
+if  [[ "$REGISTRY_CAFILE" != "" ]]; then
+  REGISTRY_CAFILE="$(base64 -w0 < $REGISTRY_CAFILE)"
+fi
+
+create_secret_manifest "$DOCKERCONFIGJSON" "$REGISTRY_ADDRESS" "$REGISTRY_PATH" "$REGISTRY_SCHEME" "$REGISTRY_CAFILE" | kubectl apply -f -
```
Comment (konstantin-axenov) [inline tools/change-registry.sh:143]: Looks good. I have a small suggestion. Why not to use `kubectl patch` command to update secret?
```
kubectl patch secret deckhouse-registry -p='{"data":{".dockerconfigjson": "dGVzdAo=", "address": "dGVzdAo=", "path": "dGVzdAo=", "scheme": "dGVzdAo="}}'
```

Comment (RomanenkoDenys) [reply tools/change-registry.sh]: > Looks good. I have a small suggestion. Why not to use `kubectl patch` command to update secret?
> 
> ```
> kubectl patch secret deckhouse-registry -p='{"data":{".dockerconfigjson": "dGVzdAo=", "address": "dGVzdAo=", "path": "dGVzdAo=", "scheme": "dGVzdAo="}}'
> ```

good sugestion, i used it )

File: docs/documentation/pages/DECKHOUSE-FAQ.md (hunk 11 above)
Comment (RomanenkoDenys) [reply docs/documentation/pages/DECKHOUSE-FAQ.md]: yep. added

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 14]
```diff
@@ -199,20 +199,24 @@ For example, you have to set `image` to `your.private.registry.com/deckhouse:v1.
...
 * Wait for bashible to apply the new settings on the master node. The bashible log on the master node (`journalctl -u bashible`) should contain the message `Configuration is in sync, nothing to do`.
 * If you want to disable Deckhouse automatic updates, remove the `releaseChannel` parameter from the `d8-system/deckhouse` ConfigMap.
 * Check if there are Pods with original registry in cluster (if there are — restart them):
 
-  ```shell
-  kubectl get pods -A -o json | jq '.items[] | select(.spec.containers[] | select((.image | contains("deckhouse.io")))) | .metadata.namespace + "\t" + .metadata.name' -r
-  ```
+```shell
+kubectl get pods -A -o json | jq '.items[] | select(.spec.containers[] | select((.image | contains("deckhouse.io")))) | .metadata.namespace + "\t" + .metadata.name' -r
+```
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:219]:
```suggestion
  ```shell
  kubectl get pods -A -o json | jq '.items[] | select(.spec.containers[] | select((.image | contains("deckhouse.io")))) | .metadata.namespace + "\t" + .metadata.name' -r
  ```
```

**Review by z9r5 — DISMISSED** [review]

**Pull request merged**